rework create-lsp7-token guide (#735)
lukso-network/docs: docs/learn/smart-contract-developers/create-lsp7-token.md

---
sidebar_label: '🌄 Create an LSP7 Token'
sidebar_position: 2
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Create an LSP7 Token

:::tip

Please check the [Getting Started](./getting-started.md) for the project setup.

:::

> ## Create a custom LSP7 Token contract

Reviewer: ```suggestion
## Create a custom LSP7 Token
```

Try to keep the headings as small but concise as possible. Otherwise, they might not fit on the side-bar.

> You will create a custom [LSP7 Digital Asset contract](../../standards/tokens/LSP7-Digital-Asset.md) that extends [`LSP7Mintable`](../../contracts/contracts/LSP7DigitalAsset/presets/LSP7Mintable.md) & [LSP7Burnable](../../contracts/contracts/LSP7DigitalAsset/extensions/LSP7Burnable.md) (to allow burning tokens). You will also pre-mint 20k tokens to the owner of the contract (the deployer).

Reviewer: ```suggestion
This guide will teach you how to create a custom [LSP7 Digital Asset](../../standards/tokens/LSP7-Digital-Asset.md) and pre-mint a certain amount of tokens to a specific address. The smart contract will use the following standardization: 

- [`LSP7Mintable`](../../contracts/contracts/LSP7DigitalAsset/presets/LSP7Mintable.md): Allow creating new assets on the smart contract
- [`LSP7Burnable`](../../contracts/contracts/LSP7DigitalAsset/extensions/LSP7Burnable.md): Allow tokens from being removed from the supply. 

You can modify the `mint()` function to adjust the amount of initially minted tokens and their receiver as described in the [LSP7 Mintable Documentation](../../contracts/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8Mintable.md#mint). 
```

I would split the sentence using bullet points so it is more apparent to the reader what a specific standardization is used for.

In the beginning, try to keep the description vague. I think explaining that he will/must mint 20k tokens to his address is unnecessary. These are just parameters that can be changed individually and discussed later down the road.

---
sidebar_label: '🌄 Create an LSP7 Token'
sidebar_position: 2
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Create an LSP7 Token

:::tip

You can learn about the project setup by checking the [Getting Started](./getting-started.md) page.

:::

## Create a custom LSP7 Token

> In this guide you will create a custom [LSP7 Digital Asset](../../standards/tokens/LSP7-Digital-Asset.md) and pre-mint a certain amount of tokens to a specific address. The smart contract will use the following standardization:

Reviewer 2: ```suggestion
In this guide you will create a custom [LSP7 Digital Asset](../../standards/tokens/LSP7-Digital-Asset.md) and pre-mint a certain amount of tokens to a specific address. To build our smart contract will use the following LSP7 preset and extension
```

> - [`LSP7Mintable`](../../contracts/contracts/LSP7DigitalAsset/presets/LSP7Mintable.md): Allow creating new assets on the smart contract

Reviewer 2: It should be lower case after colons. Also missing period.

```suggestion
- [`LSP7Mintable`](../../contracts/contracts/LSP7DigitalAsset/presets/LSP7Mintable.md): allow creating new assets on the smart contract.
```

> - [`LSP7Mintable`](../../contracts/contracts/LSP7DigitalAsset/presets/LSP7Mintable.md): Allow creating new assets on the smart contract

Reviewer 2: It should be lower case after colons. Also missing period.

```suggestion
- [`LSP7Mintable`](../../contracts/contracts/LSP7DigitalAsset/presets/LSP7Mintable.md): allow creating new assets on the smart contract.
```

npm run build
``` -->

## Deploy your LSP7 Token contract on the LUKSO Testnet

> You are now ready to deploy your contract on the [**LUKSO Testnet network**](../../networks/testnet/parameters).

Reviewer: ```suggestion
You are now ready to deploy your contract on the [**LUKSO Testnet network**](../../networks/testnet/parameters.md).
```

Don't forget `.md` on internal links. You can check if you missed it by running `npm run build` --> it should complain.

Author: i did that,this one for some reason works only without the .md, even though for most of the guides the opposite is true

> ### Create a deploy script

Reviewer: ```suggestion
### Create a deployment script
```

export default config;
```

### Create a deployment script

> Create a script to deploy the smart contract to the LUKSO Testnet network. You can either use a regular EOA (Externally Owned Account) or a Universal Profile.

Reviewer 2: So this is a script that it is intended to be run not only via the CLI, but in a frontend as well, which will trigger the Browser Extension?

Author: No, this is mainly intended to create a deployment script.

#### Deploy using a Universal Profile (Recommended)

> Deploy your contract using your Universal Profile. Make sure to have the [Universal Profile Browser Extension](/install-up-browser-extension) installed.

Reviewer 2: That's super cool, shows the capabilities of UP to deploy contracts! 🚀

:::note

The `privateKey` coming from your UP extension is the private key of the EOA that controls your UP (more information about controllers can be found in the [Key Manager](../../standards/universal-profile/lsp6-key-manager.md) page). This address will need to be funded using the [Testnet Faucet](https://faucet.testnet.lukso.network/).

:::
Create the script that will deploy the contract as your Universal Profile.

1. Create a `ethers.wallet` instance with your private key (the `signer`)
2. Load the associated UP
3. Get the bytecode of your contract
4. use `staticCall` method to get the address of the contract
5. deploy the contract

```ts title="scripts/deployUP.ts"
import hre from 'hardhat';
import { ethers } from 'hardhat';
import * as dotenv from 'dotenv';

> import LSP0ERC725Account from '@lukso/lsp-smart-contracts/artifacts/LSP0ERC725Account.json';

Reviewer 2: Suggest using UniversalProfile ABI if the idea is to instantiate your UP.

```suggestion
import UniversalProfile from '@lukso/lsp-smart-contracts/artifacts/UniversalProfile.json';
```

  });
```

Deploy the contract:

```bash
npx hardhat --network luksoTestnet run scripts/deployEOA.ts
```

> ## Check your token address on the Execution Block Explorer

Reviewer: Putting a new `h2` heading just for one sentence is unnecessary. I would suggest removing it and putting the sentence within `deployment`

> You have deployed your first LSP7 token contract on the LUKSO Testnet. You can go and check out the token on the [Execution Block Explorer](https://explorer.execution.testnet.lukso.network/) using the address you just obtained.

Reviewer: ```suggestion
:::success

After deploying your LSP7 asset on the LUKSO Testnet, you can check out the token contract on the [Execution Block Explorer](https://explorer.execution.testnet.lukso.network/). The address of the asset can be retrieved from the terminal.

:::
```

When putting it under `## Deployment`, maybe a little box can help to distinguish text sections without putting a heading. You can also just leave it as regular sentence.
